cmd/agent/app/builder.go

```diff
@@ -152,86 +146,46 @@ type SamplingServerConfiguration struct {
 	HostPort string `yaml:"hostPort" validate:"nonzero"`
 }
 
-// WithReporter adds auxiliary reporters
+// WithReporter adds a reporters. At least one reporter is required.
 func (b *Builder) WithReporter(r reporter.Reporter) *Builder {
 	b.otherReporters = append(b.otherReporters, r)
 	return b
 }
 
-// WithDiscoverer sets service discovery
-func (b *Builder) WithDiscoverer(d discovery.Discoverer) *Builder {
-	b.discoverer = d
-	return b
-}
-
-// WithDiscoveryNotifier sets service discovery notifier
-func (b *Builder) WithDiscoveryNotifier(n discovery.Notifier) *Builder {
-	b.notifier = n
-	return b
-}
-
-// WithCollectorServiceName sets collector service name
-func (b *Builder) WithCollectorServiceName(s string) *Builder {
-	b.CollectorServiceName = s
-	return b
-}
-
-// WithChannel sets tchannel channel
-func (b *Builder) WithChannel(c *tchannel.Channel) *Builder {
-	b.channel = c
-	return b
-}
-
-func (b *Builder) enableDiscovery(channel *tchannel.Channel, logger *zap.Logger) (interface{}, error) {
-	if b.discoverer == nil && b.notifier == nil {
-		return nil, nil
+func (b *Builder) createMainReporter(mFactory metrics.Factory, logger *zap.Logger) (*tchreporter.Reporter, error) {
```

Comment:
the agent always requires at least one reporter, the _main_ one (or _primary_). But nothing about that main reporter is "default", it's fully configurable.